### ISSUE TITLE
How to properly use DestroyColumnFamilyHandle?

### ISSUE THREAD

**Issue by pgoodman** [facebook/rocksdb#5006]
> Note: Please use Issues only for bug reports. For questions, discussions, feature requests, etc. post to dev group: https://www.facebook.com/groups/rocksdb.dev

### Expected behavior

No aborts / assertion failures.

### Actual behavior

Aborts / assertion failures.

### Steps to reproduce the behavior

I have two RocksDB instances, one of which uses multiple column families. It seems that there is no way to properly close these databases without one of them reporting either a double free, or one of the ColumnFamilySets freeing a ColumnFamilyData (the default one) that is marked as still being referenced.

It also surprises me that DestroyColumnFamily doesn't check if it's destroying the default column family, and in that case, null out `default_column_family_` in `DBImpl`.

I would go about trying to use multiple `Env` classes, but it seems like `PosixEnv` is hidden away in an implementation file without there being a way to create new "default-like" environments.

**Comment by pgoodman** [comment]
One sample backtrace, where I have destroyed the default column family in the instance with multiple column families, and CloseHelper is oblivious to this because the `default_cf_handle_` is non-null.

NOTE! Destroying the default column family on the instance with only one CF magically works...

```
  * frame #0: 0x00007fff58caf23e libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007fff58d65c1c libsystem_pthread.dylib`pthread_kill + 285
    frame #2: 0x00007fff58c181c9 libsystem_c.dylib`abort + 127
    frame #3: 0x00007fff58d276e2 libsystem_malloc.dylib`malloc_vreport + 545
    frame #4: 0x00007fff58d274a3 libsystem_malloc.dylib`malloc_report + 152
    frame #5: 0x000000010151d422 mu-database`rocksdb::ColumnFamilyHandleImpl::~ColumnFamilyHandleImpl(this=0x0000000105101250) at column_family.cc:49
    frame #6: 0x000000010163b0a3 mu-database`rocksdb::DBImpl::CloseHelper(this=0x0000000104892600) at db_impl.cc:447
    frame #7: 0x000000010163c35c mu-database`rocksdb::DBImpl::CloseImpl(this=0x0000000104892600) at db_impl.cc:544
    frame #8: 0x000000010165ebe3 mu-database`rocksdb::DBImpl::Close(this=0x0000000104892600) at db_impl.cc:2558
```

**Comment by pgoodman** [comment]
Another sample backtrace, this one where I haven't destroyed the default column family in the instance with multiple column families, and so it is still somehow referenced...

```
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGABRT
  * frame #0: 0x00007fff58caf23e libsystem_kernel.dylib`__pthread_kill + 10
    frame #1: 0x00007fff58d65c1c libsystem_pthread.dylib`pthread_kill + 285
    frame #2: 0x00007fff58c181c9 libsystem_c.dylib`abort + 127
    frame #3: 0x00007fff58be0868 libsystem_c.dylib`__assert_rtn + 320
    frame #4: 0x000000010152c2ae mu-database`rocksdb::ColumnFamilySet::~ColumnFamilySet(this=0x0000000104909540) at column_family.cc:1238
    frame #5: 0x000000010152c435 mu-database`rocksdb::ColumnFamilySet::~ColumnFamilySet(this=0x0000000104909540) at column_family.cc:1232
    frame #6: 0x0000000101a8c67e mu-database`rocksdb::VersionSet::~VersionSet() [inlined] std::__1::default_delete<rocksdb::ColumnFamilySet>::operator(this=0x0000000104909410, __ptr=0x0000000104909540)(rocksdb::ColumnFamilySet*) const at memory:2321
    frame #7: 0x0000000101a8c65a mu-database`rocksdb::VersionSet::~VersionSet() [inlined] std::__1::unique_ptr<rocksdb::ColumnFamilySet, std::__1::default_delete<rocksdb::ColumnFamilySet> >::reset(this=0x0000000104909410, __p=0x0000000000000000) at memory:2634
    frame #8: 0x0000000101a8c5dd mu-database`rocksdb::VersionSet::~VersionSet(this=0x0000000104909410) at version_set.cc:2776
    frame #9: 0x0000000101a8d0f5 mu-database`rocksdb::VersionSet::~VersionSet(this=0x0000000104909410) at version_set.cc:2771
    frame #10: 0x000000010163be12 mu-database`rocksdb::DBImpl::CloseHelper() [inlined] std::__1::default_delete<rocksdb::VersionSet>::operator(this=0x000000010600c228, __ptr=0x0000000104909410)(rocksdb::VersionSet*) const at memory:2321
    frame #11: 0x000000010163bdee mu-database`rocksdb::DBImpl::CloseHelper() [inlined] std::__1::unique_ptr<rocksdb::VersionSet, std::__1::default_delete<rocksdb::VersionSet> >::reset(this=0x000000010600c228, __p=0x0000000000000000) at memory:2634
    frame #12: 0x000000010163bd71 mu-database`rocksdb::DBImpl::CloseHelper(this=0x000000010600c200) at db_impl.cc:515
    frame #13: 0x000000010163c34c mu-database`rocksdb::DBImpl::CloseImpl(this=0x000000010600c200) at db_impl.cc:544
    frame #14: 0x000000010165ebd3 mu-database`rocksdb::DBImpl::Close(this=0x000000010600c200) at db_impl.cc:2558
```

**Comment by anand1976** [comment]
You need to call ```DropColumnFamily()``` followed by ```DestroyColumnFamilyHandle()```.  The default column family should not be dropped, although it is not enforced by the code.

**Comment by pgoodman** [comment]
@anand1976 Thanks for clarifying that. When I remove my calls to `DestroyColumnFamilyHandle`, I now get:
```
Assertion failed: (last_ref), function ~ColumnFamilySet, file /Users/pag/Build/multiplier-vcpkg/buildtrees/rocksdb/src/v5.17.2-53974e3cb5/db/column_family.cc, line 1238.
```

Here is what I'm doing, I've tried different variations with things commented out or not:
```c++
KeyValueStore::KeyValueEngine::~KeyValueEngine(void) {
  LOG(INFO)
      << "Shutting down key-value engine";

  auto status = rocks_db->FlushWAL(true  /* sync */);
  LOG_IF(ERROR, !status.ok() && !status.IsNotSupported())
      << "Error flushing write-ahead log: " << status.ToString();

  rocksdb::FlushOptions options;
  options.wait = true;
  status = rocks_db->Flush(options, cf_handle);
  LOG_IF(ERROR, !status.ok())
      << "Error flushing key/value store: " << status.ToString();

  LOG(INFO)
      << "Canceling background RocksDB work";
  rocksdb::CancelAllBackgroundWork(rocks_db.get(), true  /* wait */);

//  status = rocks_db->Close();
//  LOG_IF(ERROR, !status.ok())
//      << "Error closing key/value store: " << status.ToString();

  // NOTE(pag): Intentional memory leak here.
  (void) rocks_db.release();
}
```

In the above code, `rocks_db` is a `std::unique_ptr<rocksdb::DB>`. If I do `rocks_db->Close()` or if I don't intentionally leak the `DB`, then the assertion is triggered. I am not currently using the head of master (I am using whatever vcpkg points at), but when I filed the issue, I was (telling vcpkg to use `--head`).

**Comment by riversand963** [comment]
Dropping column family is kind of orthogonal to destroying a column family handle in the following ways:

1. You can destroy a column family handle without dropping a column family. This way, you just lose a "handle" to perform future operations on the said column family. The column family and its data still exist. If you close the db and reopen it, you can still get a handle to it. This applies to both default and non-default column families, except there is a trick with default column family: you should not call `DB::DestroyColumnFamilyHandle()` with the column family handle returned by `DB::DefaultColumnFamily()`. #9347 adds a check for this case.
2. You can drop a column family without destroying a handle to it. Before you destroy the handle, you can still read from it, but you probably cannot write to it or flush it (I didn't double check). After you destroy the handle, the db will at some point delete its data and remove the column family from `ColumnFamilySet`.

To sum up, you can call either method independently depending on your use case.

I am going to close this one once the PR is merged.
